### DISCUSSION TITLE
How to make Fastapi put the stacktrace in the body when handling 500 error

### DISCUSSION THREAD

**Discussion by allan-simon** [fastapi/fastapi#7519] [Questions] [answered]
### First check

* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description

During the development cycle, especially when running on a remote machine (QA environment etc.) it's pretty convenient to have the traceback directly in the body of a 500 response so that one does not have to connect to the machine to get the traceback.

Of course it's something that one should opt-in , as you don't want to have that in production

**Comment by phy25** [comment]
You could start with https://fastapi.tiangolo.com/tutorial/handling-errors/#install-custom-exception-handlers with a call to a function in `traceback` module.

**Comment by allan-simon** [comment]
ah yeah right , stupid me, thanks for the pointer, I will try to come up with something and put it here in case somebody wonders the same thing

**Comment by allan-simon** [comment, marked as answer]
ok I've hacked this and it seems to work pretty well 

```py
if DISPLAY_TRACEBACK_ON_500:

    @app.exception_handler(Exception)
    async def debug_exception_handler(request: Request, exc: Exception):
        import traceback

        return Response(
            content="".join(
                traceback.format_exception(
                    etype=type(exc), value=exc, tb=exc.__traceback__
                )
            )
        )


```

**Reply by kubosuke** [reply]
mine:

```python
@app.exception_handler(Exception)
async def debug_exception_handler(_: Request, exc: Exception):
    return Response(content="".join(traceback.TracebackException.from_exception(exc).format()))
```

**Comment by phy25** [comment]
You may want to close the issue if it works for you :)

**Comment by allan-simon** [comment]
haha sure

**Comment by tiangolo** [comment]
Thanks for the help here @phy25 ! :cake: :bow: 

Thanks @allan-simon for coming back to close it :+1:

**Comment by retnikt** [comment]
Starlette has a built-in server exception middleware - use `FastAPI(debug=True)` to enable it

**Comment by mhulo** [comment]
> Starlette has a built-in server exception middleware - use `FastAPI(debug=True)` to enable it

The fact that you can set debug=true is mentioned in several places, but not where to or how to. I finally saw your reply after 2hrs of googling. Would be great to get this into the fastapi docs.

**Comment by perkfly** [comment]
I can get FastAPI to print stack trace with `debug=True`, but how to make FastAPI print the trackback on command line?

**Comment by mhulo** [comment]
Surprisingly (to me) its late 2021 and the starlette debug traceback response whilst pretty, still does not show the local vars - as other frameworks do (eg. django).. and this was essential to me migrating some existing projects to fastApi.

My crude workaround is as follows;

```#global exception handler with traceback
if (custom_debug == True):
  @app.exception_handler(Exception)
  async def debug_exception_handler(request: Request, exc: Exception):

      import stackprinter
      sp = stackprinter.format(reverse=True)

      sp = sp.replace('<', '&lt; ')
      sp = sp.replace('>', '&gt; ')
      sp = sp.replace('--\n\nFile', '--|||File')
      sp = sp.replace('..\n\nFile', '..|||File')
      sp_arr = sp.split('|||')
     
      main = ''
      for i in sp_arr:
        main += '<div class="sect"><pre>' + i + '</pre></div><br>\n'

      resp =  '''
        <html>
          <head>
            <style type="text/css">
              body { border: 1px #038BB8 solid; }
              .hdr { border-bottom: 20px #038BB8 solid; padding-top: 1px; padding-left: 10px; }
              .sect { padding: 10px; background: #E4F4FD; border: 1px #c7dce8 solid; border-width: 1px 0px; }
            </style>
          </head>
          <body>
            <div class="hdr">
              <h2>500 Server Error:</h2>
            </div>
            <br>
          '''
      resp += main
      resp += '<body></html>'
        
      return HTMLResponse(content=resp)

**Comment by tibbe** [comment]
What are the pros/cons of this approach:

```py
import logging
from typing import Callable

from fastapi import FastAPI, Response
from fastapi.exceptions import HTTPException, RequestValidationError
from fastapi.routing import APIRoute
from starlette.requests import Request

class ServerErrorLoggingRoute(APIRoute):
    '''Log uncaught exceptions.'''

    def get_route_handler(self) -> Callable:
        original_route_handler = super().get_route_handler()

        async def custom_route_handler(request: Request) -> Response:
            try:
                return await original_route_handler(request)
            # These exceptions are used for control flow rather to indicate a
            # server error so we don't log these.
            except RequestValidationError:
                raise
            except HTTPException:
                raise
            except Exception:
                logging.exception('Uncaught exception')
                raise

        return custom_route_handler

app = FastAPI()
app.router.route_class = ServerErrorLoggingRoute
```
